### ISSUE TITLE
[SDK 44] Attempt to invoke interface method 'expo.modules.interfaces.facedetector.FaceDetectorInterface' on a null object reference

### ISSUE THREAD

**Issue by ajacquierbret** [expo/expo#15547]
### Summary

On Android, when mounting the Camera component from `expo-camera`, an error is thrown. Dismissing the error doesn't prevent the camera from being mounted and detecting barcodes. Installing the `expo-face-detector` package prevents the error from being thrown.

Error logs from Flipper :

```
Exception in native call
java.lang.NullPointerException: Attempt to invoke interface method 'expo.modules.interfaces.facedetector.FaceDetectorInterface expo.modules.interfaces.facedetector.FaceDetectorProviderInterface.createFaceDetectorWithContext(android.content.Context)' on a null object reference
	at expo.modules.camera.ExpoCameraView.onHostResume(ExpoCameraView.kt:214)
	at expo.modules.adapters.react.services.UIManagerModuleWrapper$3.onHostResume(UIManagerModuleWrapper.java:127)
	at com.facebook.react.bridge.ReactContext.onHostResume(ReactContext.java:240)
	at com.facebook.react.ReactInstanceManager.moveToResumedLifecycleState(ReactInstanceManager.java:719)
	at com.facebook.react.ReactInstanceManager.onHostResume(ReactInstanceManager.java:623)
	at com.facebook.react.ReactInstanceManager.onHostResume(ReactInstanceManager.java:580)
	at com.facebook.react.ReactDelegate.onHostResume(ReactDelegate.java:53)
	at com.facebook.react.ReactActivityDelegate.onResume(ReactActivityDelegate.java:100)
	at java.lang.reflect.Method.invoke(Native Method)
	at expo.modules.ReactActivityDelegateWrapper.invokeDelegateMethod(ReactActivityDelegateWrapper.kt:157)
	at expo.modules.ReactActivityDelegateWrapper.onResume(ReactActivityDelegateWrapper.kt:81)
	at com.facebook.react.ReactActivity.onResume(ReactActivity.java:58)
	at android.app.Instrumentation.callActivityOnResume(Instrumentation.java:1456)
	at android.app.Activity.performResume(Activity.java:8135)
	at android.app.ActivityThread.performResumeActivity(ActivityThread.java:4428)
	at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:4470)
	at android.app.servertransaction.ResumeActivityItem.execute(ResumeActivityItem.java:52)
	at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:176)
	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2066)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loop(Looper.java:223)
	at android.app.ActivityThread.main(ActivityThread.java:7660)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
```

### Managed or bare workflow? If you have `ios/` or `android/` directories in your project, the answer is bare!

bare

### What platform(s) does this occur on?

Android

### SDK Version (managed workflow only)

44.0.0-beta.1

### Environment

```
Expo CLI 5.0.0 environment info:
    System:
      OS: macOS 11.4
      Shell: 5.8 - /bin/zsh
    Binaries:
      Node: 14.18.1 - ~/.asdf/installs/nodejs/14.18.1/bin/node
      Yarn: 1.22.10 - /usr/local/bin/yarn
      npm: 6.14.15 - ~/.asdf/installs/nodejs/14.18.1/bin/npm
      Watchman: 2021.11.15.00 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.11.2 - /opt/homebrew/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 21.2, iOS 15.2, macOS 12.1, tvOS 15.2, watchOS 8.3
      Android SDK:
        API Levels: 29, 30, 31
        Build Tools: 29.0.2, 29.0.3, 30.0.2, 30.0.3
        System Images: android-29 | Google APIs ARM 64 v8a, android-30 | Google APIs ARM 64 v8a, android-30 | Google APIs Intel x86 Atom, android-31 | Google APIs ARM 64 v8a, android-32 | Google APIs ARM 64 v8a, android-32 | Google APIs Intel x86 Atom_64
    IDEs:
      Xcode: 13.2/13C5081f - /usr/bin/xcodebuild
    npmPackages:
      babel-preset-expo: 9.0.1 => 9.0.1
      expo: ^44.0.0-beta.1 => 44.0.0-beta.1
      react: 17.0.1 => 17.0.1
      react-dom: 17.0.1 => 17.0.1
      react-native: 0.64.3 => 0.64.3
      react-native-web: 0.17.1 => 0.17.1
    Expo Workflow: bare
```

### Reproducible demo

https://github.com/ajacquierbret/expo-issue-15547

**Comment by brentvatne** [comment]
> Since the Expo SDK 44 isn't available yet on Snack, I can't share a repro.

please share a github repo

**Comment by ajacquierbret** [comment]
@brentvatne Sorry for the delay, here's a bare workflow minimal GitHub repro of the issue : https://github.com/ajacquierbret/expo-issue-15547

Steps to reproduce the issue :

Run the project as-is using `yarn && expo run:android --device` and select a real device.
You'll encounter this error when the camera component mounts :

```
Exception in native call
java.lang.NullPointerException: Attempt to invoke interface method 'expo.modules.interfaces.barcodescanner.BarCodeScannerInterface expo.modules.interfaces.barcodescanner.BarCodeScannerProviderInterface.createBarCodeDetectorWithContext(android.content.Context)' on a null object reference
	at expo.modules.camera.ExpoCameraView.initBarCodeScanner(ExpoCameraView.kt:182)
	at expo.modules.camera.ExpoCameraView.<init>(ExpoCameraView.kt:273)
	at expo.modules.camera.CameraViewManager.createViewInstance(CameraViewManager.kt:46)
	at expo.modules.camera.CameraViewManager.createViewInstance(CameraViewManager.kt:15)
	at expo.modules.adapters.react.views.ViewGroupManagerAdapter.createViewInstance(ViewGroupManagerAdapter.java:27)
	at expo.modules.adapters.react.views.ViewGroupManagerAdapter.createViewInstance(ViewGroupManagerAdapter.java:18)
	at com.facebook.react.uimanager.ViewManager.createViewInstance(ViewManager.java:139)
	at com.facebook.react.uimanager.ViewManager.createView(ViewManager.java:83)
	at com.facebook.react.uimanager.NativeViewHierarchyManager.createView(NativeViewHierarchyManager.java:281)
	at com.facebook.react.uimanager.UIViewOperationQueue$CreateViewOperation.execute(UIViewOperationQueue.java:188)
	at com.facebook.react.uimanager.UIViewOperationQueue$DispatchUIFrameCallback.dispatchPendingNonBatchedOperations(UIViewOperationQueue.java:1103)
	at com.facebook.react.uimanager.UIViewOperationQueue$DispatchUIFrameCallback.doFrameGuarded(UIViewOperationQueue.java:1074)
	at com.facebook.react.uimanager.GuardedFrameCallback.doFrame(GuardedFrameCallback.java:29)
	at com.facebook.react.modules.core.ReactChoreographer$ReactChoreographerDispatcher.doFrame(ReactChoreographer.java:175)
	at com.facebook.react.modules.core.ChoreographerCompat$FrameCallback$1.doFrame(ChoreographerCompat.java:85)
	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:970)
	at android.view.Choreographer.doCallbacks(Choreographer.java:796)
	at android.view.Choreographer.doFrame(Choreographer.java:727)
	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:957)
	at android.os.Handler.handleCallback(Handler.java:938)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loop(Looper.java:223)
	at android.app.ActivityThread.main(ActivityThread.java:7660)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)

```

Then, install the `expo-barcode-scanner` package using `expo install expo-barcode-scanner package`.
Re-run the project using `expo run:android --device`, and you'll see this error showing up :

```
FATAL EXCEPTION: main
Process: com.silancey.expoissue15547, PID: 19483
java.lang.NullPointerException: Attempt to invoke interface method 'expo.modules.interfaces.facedetector.FaceDetectorInterface expo.modules.interfaces.facedetector.FaceDetectorProviderInterface.createFaceDetectorWithContext(android.content.Context)' on a null object reference
	at expo.modules.camera.ExpoCameraView.onHostResume(ExpoCameraView.kt:214)
	at expo.modules.adapters.react.services.UIManagerModuleWrapper$3.onHostResume(UIManagerModuleWrapper.java:127)
	at com.facebook.react.bridge.ReactContext$1.run(ReactContext.java:200)
	at android.os.Handler.handleCallback(Handler.java:938)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at com.facebook.react.bridge.queue.MessageQueueThreadHandler.dispatchMessage(MessageQueueThreadHandler.java:27)
	at android.os.Looper.loop(Looper.java:223)
	at android.app.ActivityThread.main(ActivityThread.java:7660)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)

```

In order to run the project without errors, install the `expo-face-detector` package using `expo install expo-face-detector` and re-run the project using `expo run:android --device`.

**Comment by brentvatne** [comment]
thank you for the info!

**Comment by kevgrig** [comment]
In my case, it didn't just cause a dismissable error but a full app crash. Thank you @ajacquierbret for the work-around of `expo install expo-barcode-scanner expo-face-detector` which works.

```
12-26 13:13:11.480 12758 12758 D AndroidRuntime: Shutting down VM
12-26 13:13:11.481 12758 12758 E AndroidRuntime: FATAL EXCEPTION: main
12-26 13:13:11.481 12758 12758 E AndroidRuntime: Process: [...], PID: 12758
12-26 13:13:11.481 12758 12758 E AndroidRuntime: java.lang.NullPointerException: Attempt to invoke interface method 'expo.modules.interfaces.barcodescanner.BarCodeScannerInterface expo.modules.interfaces.barcodescanner.BarCodeScannerProviderInterface.createBarCodeDetectorWithContext(android.content.Context)' on a null object reference
12-26 13:13:11.481 12758 12758 E AndroidRuntime:        at expo.modules.camera.ExpoCameraView.initBarCodeScanner(ExpoCameraView.kt:182)
12-26 13:13:11.481 12758 12758 E AndroidRuntime:        at expo.modules.camera.ExpoCameraView.<init>(ExpoCameraView.kt:273)
12-26 13:13:11.481 12758 12758 E AndroidRuntime:        at expo.modules.camera.CameraViewManager.createViewInstance(CameraViewManager.kt:46)
12-26 13:13:11.481 12758 12758 E AndroidRuntime:        at expo.modules.camera.CameraViewManager.createViewInstance(CameraViewManager.kt:15)
12-26 13:13:11.481 12758 12758 E AndroidRuntime:        at expo.modules.adapters.react.views.ViewGroupManagerAdapter.createViewInstance(ViewGroupManagerAdapter.java:27)
[...]
12-26 13:13:11.483  1828  7752 W ActivityTaskManager:   Force finishing activity [...]/.MainActivity
12-26 13:13:11.492 12758 12758 I Process : Sending signal. PID: 12758 SIG: 9
12-26 13:13:11.523  1828  4416 I WindowManager: WIN DEATH: Window{10b3942 u0 [...]/[...].MainActivity}
12-26 13:13:11.524  1160  1160 I Zygote  : Process 12758 exited due to signal 9 (Killed)
12-26 13:13:11.524  1828  1910 I libprocessgroup: Successfully killed process cgroup uid 10374 pid 12758 in 0ms
```

**Comment by yonitou** [comment]
I'm having the same issue using a managed workflow on SDK 44. 

Here's the sentry issue : 

`NullPointerException: Attempt to invoke interface method 'expo.modules.interfaces.facedetector.FaceDetectorInterface expo.modules.interfaces.facedetector.FaceDetectorProviderInterface.createFaceDetectorWithContext(android.content.Context)' on a null object reference`

**Comment by eeynard** [comment]
Hi guys ! I have the same type of issue but slightly different `Attempt to invoke interface method 'expo.modules.interfaces.barcodescanner.BarCodeScannerInterface expo.modules.interfaces.barcodescanner.BarCodeScannerProviderInterface.createBarCodeDetectorWithContext(android.content.Context)' on a null object reference`

ℹ️  After downgrading expo-camera from 12.1.0 to 12.0.3, the issue does not seem to show anymore.

**Comment by Yonom** [comment]
It seems like `expo-camera` v12.1 requires `expo-barcode-scanner` and `expo-face-detector` to work.

Running the following commands fixes the problem for me (but adds two unused dependencies)
```
expo install expo-barcode-scanner
expo install expo-face-detector
```

**Comment by NorthBlue333** [comment]
Also got this issue.
I already use `expo-barcode-scanner`, but I do not use `expo-face-detector`. Adding this last one adds 20 MB to the `.aab` (android) production build (originally the build was `39.8 MB`, `56.5 MB` after adding it).
I ended up downgrading manually the package as @eeynard suggested.

**Comment by yonitou** [comment]
On my side, the app crashes on Android when, on a screen using expo-camera, sending the app to background and then foreground.

<img width="764" alt="Capture d’écran 2022-01-07 à 11 51 32" src="https://user-images.githubusercontent.com/61981305/148533413-7eaba2e8-5ff7-467e-958a-7348fc28848c.png">

**Comment by instakash** [comment]
I have the same problem. I upgraded from SDK 43 to 44 and the only screen I have using the expo camera now crashes on android devices.

![Screen Shot 2022-01-08 at 15 29 47](https://user-images.githubusercontent.com/77989503/148659012-1410f18c-1e4c-4f4b-aea6-e81e1d698b14.png)

**Comment by islamouzou** [comment]
@brentvatne these are my logs from playstore  : i only use expo-camera no barcode scanner

java.lang.NullPointerException: 
  at expo.modules.camera.ExpoCameraView.initBarCodeScanner (ExpoCameraView.kt:182)
  at expo.modules.camera.ExpoCameraView.<init> (ExpoCameraView.kt:273)
  at expo.modules.camera.CameraViewManager.createViewInstance (CameraViewManager.kt:46)
  at expo.modules.camera.CameraViewManager.createViewInstance (CameraViewManager.kt:15)
  at expo.modules.adapters.react.views.ViewGroupManagerAdapter.createViewInstance (ViewGroupManagerAdapter.java:27)
  at expo.modules.adapters.react.views.ViewGroupManagerAdapter.createViewInstance (ViewGroupManagerAdapter.java:18)
  at com.facebook.react.uimanager.ViewManager.createViewInstance (ViewManager.java:139)
  at com.facebook.react.uimanager.ViewManager.createView (ViewManager.java:83)
  at com.facebook.react.uimanager.NativeViewHierarchyManager.createView (NativeViewHierarchyManager.java:281)
  at com.facebook.react.uimanager.UIViewOperationQueue$CreateViewOperation.execute (UIViewOperationQueue.java:188)
  at com.facebook.react.uimanager.UIViewOperationQueue$DispatchUIFrameCallback.dispatchPendingNonBatchedOperations (UIViewOperationQueue.java:1103)
  at com.facebook.react.uimanager.UIViewOperationQueue$DispatchUIFrameCallback.doFrameGuarded (UIViewOperationQueue.java:1074)
  at com.facebook.react.uimanager.GuardedFrameCallback.doFrame (GuardedFrameCallback.java:29)
  at com.facebook.react.modules.core.ReactChoreographer$ReactChoreographerDispatcher.doFrame (ReactChoreographer.java:175)
  at com.facebook.react.modules.core.ChoreographerCompat$FrameCallback$1.doFrame (ChoreographerCompat.java:85)
  at android.view.Choreographer$CallbackRecord.run (Choreographer.java:1063)
  at android.view.Choreographer.doCallbacks (Choreographer.java:889)
  at android.view.Choreographer.doFrame (Choreographer.java:812)
  at android.view.Choreographer$FrameDisplayEventReceiver.run (Choreographer.java:1050)
  at android.os.Handler.handleCallback (Handler.java:883)
  at android.os.Handler.dispatchMessage (Handler.java:100)
  at android.os.Looper.loop (Looper.java:237)
  at android.app.ActivityThread.main (ActivityThread.java:7824)
  at java.lang.reflect.Method.invoke (Native Method)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:512)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1027)

**Comment by paulocappa** [comment]
Up! I have the same problem here

**Comment by n0mer** [comment]
> It seems like `expo-camera` v12.1 requires `expo-barcode-scanner` and `expo-face-detector` to work.
> 
> Running the following commands fixes the problem for me (but adds two unused dependencies)
> 
> ```
> expo install expo-barcode-scanner
> expo install expo-face-detector
> ```

Strange that those packages are not mentioned in https://github.com/expo/expo/blob/master/packages/expo-image-picker/package.json

**Comment by raRaRa** [comment]
Same issue here, downgraded `expo-camera` to 12.0.3 fixed it.

**Comment by taofik-adeniyi** [comment]
Having same issue here followed all the steps listed above but still got the same error below
Attempt to invoke interface method expo.modules.interfaces.barcodescanner.BarCodeScannerInterface expo.modules.interfaces.barcodescanner.BarCodeScannerProviderInterface.createBarCodeDetectorWithContext(android.content.Context)` on a null object reference

React native cli app

**Comment by brentvatne** [comment]
fix deployed in expo-camera@12.1.2

**Comment by NorthBlue333** [comment]
Hi @brentvatne 
As @yonitou specified, there is an issue when bringing the app to the foreground (on Android at least). I think it can be reproduced with the same demo. You have to open the camera view, go to home screen, and bring the app back to foreground.
